test/integration/auth/auth.prose.test.ts

```diff
@@ -0,0 +1,416 @@
+        async function () {
+          const options = {
+            auth: {
+              username: userMap.sha256.username,
+              password: userMap.sha256.password
+            },
+            authSource: 'admin',
+            authMechanism: 'SCRAM-SHA-1'
+          };
+
+          const client = this.configuration.newClient({}, options);
+          const error = await client
+            .db('test')
+            .stats()
+            .catch(e => e);
+          expect(error.message).to.match(/Authentication failed|SCRAM/);
+          await client.close();
+        }
+      );
+
+      it(
+        'fails to connect if incorrect auth mechanism (SCRAM-SHA-256) is explicitly specified',
+        metadata,
+        async function () {
+          const options = {
+            auth: {
+              username: userMap.sha1.username,
+              password: userMap.sha1.password
+            },
+            authSource: 'admin',
+            authMechanism: 'SCRAM-SHA-256'
+          };
+
+          const client = this.configuration.newClient({}, options);
+          const error = await client
+            .db('test')
+            .stats()
+            .catch(e => e);
+          expect(error.message).to.match(/Authentication failed|SCRAM/);
+          await client.close();
+        }
+      );
+
+      /*
+       * Step 3
+       * For a non-existent username, verify that not specifying a mechanism when
+       * connecting fails with the same error type that would occur with a correct
+       * username but incorrect password or mechanism.  (Because negotiation with a
+       * non-existent user name at one point during server development caused a
+       * handshake error, we want to verify this is seen by users as similar to other
+       * authentication errors, not as a network or database command error on the ``hello``
+       * or legacy hello commands themselves.)
+       */
+      it(
+        'fails for a nonexistent username with same error type as bad password',
+        metadata,
+        async function () {
+          const noUsernameOptions = {
+            auth: {
+              password: 'pencil'
+            },
+            authSource: 'admin'
+          };
+
+          const badPasswordOptions = {
+            auth: {
+              username: 'both',
+              password: 'pencil'
+            },
+            authSource: 'admin'
+          };
+
+          try {
+            this.configuration.newClient({}, noUsernameOptions);
+            expect.fail(
+              'Creating a new client with a password and no username must fail validation.'
+            );
+          } catch (noUserError) {
+            // NOTE: This prose test fails Node's validation of the credentials object - a username
+            // MUST always be provided but it satisfies the test requirement of not getting a
+            // network or command error on the handshake.
+            expect(noUserError).to.match(/username/);
+          }
```

Comment:
The expect.fail has "username" in the string but I think our intention is to seek a username substring from a driver error correct?

Reply:
Yes we are expecting the options validation to throw saying the username is missing. Was there a change request here?

Reply:
The `.match(/username/)` assertion will _always_ pass because the `expect.fail` _also_ satisfies `/username/`. Either we need to change the string in `expect.fail` or restructure this to store the error into a variable and assert outside of the `try/catch`

Reply:
Ahhh. Now I see. :)

Reply:
Fail message changed.